### PR TITLE
ARROW-10520: [C++][R] Implement add/remove/replace for RecordBatch

### PR DESCRIPTION
**Pull request by nealrichardson**
In C++, RecordBatch had AddColumn and RemoveColumn but no SetColumn; Table had all three. This adds the missing RecordBatch method and adds R bindings.

### CODE REVIEW

**Review by pitrou — COMMENTED** [review]
LGTM, just one nit

File: cpp/src/arrow/record_batch.cc [hunk 1]
```diff
@@ -104,6 +104,27 @@ class SimpleRecordBatch : public RecordBatch {
...
   }
 
+  Result<std::shared_ptr<RecordBatch>> SetColumn(
+      int i, const std::shared_ptr<Field>& field,
+      const std::shared_ptr<Array>& column) const override {
+    ARROW_CHECK(field != nullptr);
+    ARROW_CHECK(column != nullptr);
+
+    if (!field->type()->Equals(column->type())) {
+      return Status::Invalid("Column data type ", field->type()->name(),
```
[the remark this answers is not in the document]
Comment (nealrichardson) [reply cpp/src/arrow/record_batch.cc:114]: Sure. This is just copied from L92 above; should I change that one too?

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/record_batch.cc:114]: Ah, that would be good. You may need to fix some tests too :-)

[the remark this answers is not in the document]
Comment (nealrichardson) [reply cpp/src/arrow/record_batch.cc]: I changed 3 in this file, PTAL

**Review by pitrou — APPROVED** [review]
+1, thank you. Let's just wait for CI before merging.

**Pull request closed without merging**